src/java/org/apache/cassandra/db/compaction/ShardManagerNodeAware.java

```diff
@@ -0,0 +1,280 @@
+import org.apache.cassandra.dht.IPartitioner;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.dht.tokenallocator.TokenAllocation;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.sstable.format.SSTableWriter;
+import org.apache.cassandra.locator.AbstractReplicationStrategy;
+import org.apache.cassandra.locator.TokenMetadata;
+
+/**
+ * A shard manager implementation that accepts token-allocator-generated-tokens and splits along them to ensure that
+ * current and future states of the cluster will have sstables within shards, not across them, for sufficiently high
+ * levels of compaction, which allows nodes to trivially own complete sstables for sufficiently high levels of
+ * compaction.
+ *
+ * If there are not yet enough tokens allocated, use the {@link org.apache.cassandra.dht.tokenallocator.TokenAllocator}
+ * to allocate more tokens to split along. The key to this implementation is utilizing the same algorithm to allocate
+ * tokens to nodes and to split ranges for higher levels of compaction.
+ */
+    // I haven't figured out yet whether the interesting part of this class is the fact that we use the token allocator
+    // to find higher level splits or if it is the node awareness. Is it possible to remove the node awareness and keep
+    // the allocator's logic or do we need both?
+// TODO should we extend ShardManagerDiskAware?
+
+public class ShardManagerNodeAware implements ShardManager
+{
+    public static final Token[] TOKENS = new Token[0];
+    private final AbstractReplicationStrategy rs;
+    private final TokenMetadata tokenMetadata;
+
+    public ShardManagerNodeAware(AbstractReplicationStrategy rs)
+    {
+        this.rs = rs;
+        this.tokenMetadata = rs.getTokenMetadata();
+    }
+
+    @Override
+    public double rangeSpanned(Range<Token> tableRange)
+    {
+        return tableRange.left.size(tableRange.right);
+    }
+
+    @Override
+    public double localSpaceCoverage()
+    {
+        // At the moment, this is global, so it covers the whole range. Might not be right though.
+        return 1;
+    }
+
+    @Override
+    public double shardSetCoverage()
+    {
+        // For now there are no disks defined, so this is the same as localSpaceCoverage
+        return 1;
+    }
+
+    @Override
+    public ShardTracker boundaries(int shardCount)
+    {
+        var splitPointCount = shardCount - 1;
+        // TODO is it safe to get tokens here and then endpoints later without synchronization?
+        var sortedTokens = tokenMetadata.sortedTokens();
+        if (splitPointCount > sortedTokens.size())
+        {
+            // Need to allocate tokens within node boundaries.
+            var endpoints = tokenMetadata.getAllEndpoints();
+            double addititionalSplits = splitPointCount - sortedTokens.size();
+            var splitPointsPerNode = (int) Math.ceil(addititionalSplits / endpoints.size());
```

Comment:
I was thinking about it the wrong way. I was focused on finding the right splits for the existing nodes instead of just adding fake new nodes until we have enough tokens. This seems more straight forward.

Reply:
> We can then flatten the generated token list and truncate it to exactly the required number of extra tokens, then concatenate it at the end of the original tokens and sort (i.e. prefer just choosing the first `additionalSplits` many from the new instead of using the selection scheme below).

I would have thought the selection scheme would give us better data distribution, as opposed to truncating the list of new tokens. Also, if we truncate the list, does that present issues for ensuring that higher levels of UCS have the same splits as lower levels?

Reply:
I just pushed up a new commit with a proposed implementation. I plan to write tests for it tomorrow.

Reply:
> if we truncate the list, does that present issues for ensuring that higher levels of UCS have the same splits as lower levels?

No, it does not, because the same tokens will be generated again when we generate the higher-count levels (in other words, we can cache the smaller generation round (pre-truncation) and use it to save some work when generating the bigger).

> I would have thought the selection scheme would give us better data distribution, as opposed to truncating the list of new tokens

The token allocation gives the biggest-impact-token first, which is also the one that splits the current biggest token range so taking them in order should be a good enough choice and we can save the effort. Also, I'm not sure looking for closest to even split can't cause bigger oddities.